src/terminal/input/terminalInput.cpp

```diff
@@ -515,169 +162,442 @@ TerminalInput::OutputType TerminalInput::HandleKey(const INPUT_RECORD& event)
+
+    // Using the control key state that we calculated above, combined with the
+    // virtual key code, we've got a unique identifier for the key combination
+    // that we can lookup in our map of predefined key sequences.
+    auto keyCombo = virtualKeyCode;
+    keyCombo += (ctrlIsReallyPressed ? Ctrl : Unmodified);
+    keyCombo += (altIsPressed ? Alt : Unmodified);
+    keyCombo += (shiftIsPressed ? Shift : Unmodified);
+    keyCombo += (enhancedReturnKey ? Enhanced : Unmodified);
+    auto keySequence = _keyMap[keyCombo];
+    if (!keySequence.empty())
     {
-        const auto ch = keyEvent.uChar.UnicodeChar;
-        const auto vkey = keyEvent.wVirtualKeyCode;
-
-        // For Alt+Ctrl+Key messages uChar.UnicodeChar usually returns 0.
-        // Luckily the numerical values of the ASCII characters and virtual key codes
-        // of <Space> and A-Z, as used below, are numerically identical.
-        // -> Get the char from the virtual key if it's 0.
-        const auto ctrlAltChar = keyEvent.uChar.UnicodeChar != 0 ? keyEvent.uChar.UnicodeChar : keyEvent.wVirtualKeyCode;
-
-        // Alt+Ctrl acts as a substitute for AltGr on Windows.
-        // For instance using a German keyboard both AltGr+< and Alt+Ctrl+< produce a | (pipe) character.
-        // The below condition primitively ensures that we allow all common Alt+Ctrl combinations
-        // while preserving most of the functionality of Alt+Ctrl as a substitute for AltGr.
-        if (ctrlAltChar == UNICODE_SPACE || (ctrlAltChar > 0x40 && ctrlAltChar <= 0x5A))
+        return keySequence;
+    }
+
+    // If it's not in the key map, we'll use the UnicodeChar, if provided.
+    if (unicodeChar != 0)
+    {
+        // In the case of an AltGr key, we may still need to apply a Ctrl
+        // modifier to the char, either because both Ctrl keys were pressed,
+        // or we got a LeftCtrl that was distinctly separate from the RightAlt.
+        const auto bothCtrlsArePressed = WI_AreAllFlagsSet(controlKeyState, CTRL_PRESSED);
+        const auto rightAltIsPressed = WI_IsFlagSet(controlKeyState, RIGHT_ALT_PRESSED);
+        if (altGrIsPressed && (bothCtrlsArePressed || (rightAltIsPressed && leftCtrlIsReallyPressed)))
         {
-            // Pressing the control key causes all bits but the 5 least
-            // significant ones to be zeroed out (when using ASCII).
-            return _makeEscapedOutput(ctrlAltChar & 0b11111);
+            unicodeChar = _makeCtrlChar(unicodeChar);
         }
-
-        // Currently, when we're called with Alt+Ctrl+@, ch will be 0, since Ctrl+@ equals a null byte.
-        // VkKeyScanW(0) in turn returns the vkey for the null character (ASCII @).
-        // -> Use the vkey to determine if Ctrl+@ is being pressed and produce ^[^@.
-        if (ch == UNICODE_NULL && vkey == LOBYTE(OneCoreSafeVkKeyScanW(0)))
+        auto charSequence = _makeCharOutput(unicodeChar);
+        if (!charSequence.empty())

```

Comment:
By moving the surrogate pair check up in this function we could make it so that `_makeCharOutput` never returns an empty string.

Reply:
I'm not sure what you meant here. Did you want to move the whole `if til::is_leading_surrogate(ch)) _leadingSurrogate.emplace(ch)` test out of the `_makeCharOutput` function? Because this isn't the only place that's being called. It's also used for the `VK_PACKET` and 0 virtual key events (although I'm not actually sure we should be delaying the surrogate pairs in that case).

Reply:
What I meant is that we could consider moving the leading-surrogate check right past the `if (!keyEvent.bKeyDown)` branch. It would then store any leading-surrogate into a member variable and could then early-return. That is:
```cpp
class TerminalInput
{
    // no need for an optional
    wchar_t _leadingSurrogate = 0;
};

TerminalInput::OutputType TerminalInput::HandleKey(const INPUT_RECORD& event)
{
    // ...

    if (!keyEvent.bKeyDown)
    {
        // ...
    }

    // Unpaired surrogates are no good -> early return.
    if (til::is_leading_surrogate(unicodeChar))
    {
        _leadingSurrogate = unicodeChar;
        return MakeOutput({});
    }
    // Using a scope_exit ensures that a previous leading surrogate
    // is forgotten even if the KEY_EVENT that followed is 
    const auto leadingSurrogateReset = wil::scope_exit([&]() {
        _leadingSurrogate = 0;
    });

    // ...
}

TerminalInput::StringType TerminalInput::_makeCharOutput(const wchar_t ch)
{
    StringType str;

    if (_leadingSurrogate && til::is_trailing_surrogate(ch))
    {
        str.push_back(_leadingSurrogate);
    }

    str.push_back(ch);
    return str;
}
```

That way `_makeCharOutput` never returns an empty string which may allow you to simplify the `if (unicodeChar != 0)` branch.

However, there's one thing that made me wonder a bit now. The `VK_PACKET` handler doesn't deal with Alt key combinations. Why's that? Shouldn't it use the exact same code as the `if (unicodeChar != 0)` branch? Because if it did, you could abstract it away and then call that function directly when encountering a trailing surrogate key, since no character from a higher Unicode plane should ever hit any of the VK_* handlers, etc., anyways (I think?).

---

BTW: It may be worth introducing a `_returnNoOutput` function or similar, since we now have 7 instances of `MakeOutput({})`, but it's conceptually the same thing every time.

Reply:
OK. That makes sense. I've incorporated those changes now. It does simplify things. Thanks.

As for `VK_PACKET`, my understanding was that it serves as a kind of "pass through" mode, when you just want to send a raw unicode character as input. It will typically come from something like an emoji picker, or a `SendInput` call with `KEYEVENTF_UNICODE`.

It's not an actual keypress, so I don't think it's supposed to be affected by modifiers. If you wanted to simulate a VT Alt key with `VK_PACKET` you'd literally send through the sequence of unicode characters that make up that Alt key sequence.